lib/install/rust.js

```diff
@@ -0,0 +1,559 @@
+            }
+
+            // Write out CHECKSUM file.
+            fs.writeFileSync(path.join(destdir.path, 'CHECKSUM'), checksum);
+
+            try {
+              // Remove the old SDK directory.
+              fs.removeSync(root);
+              // Move temporary directory to target destination.
+              fs.move(destdir.path, root, (err) => {
+                if (err) {
+                  // Cleanup temp dir.
+                  destdir.cleanup();
+                  reject(err);
+                } else {
+                  resolve();
+                }
+              });
+            } catch (e) {
+              // Cleanup temp dir.
+              destdir.cleanup();
+              reject(e);
+            }
+          })
+          .on('error', (err) => {
+            destdir.cleanup();
+            reject(err);
+          });
+      });
+    });
+}
+
+function extractTools(checksumVerify, filename, sdkStream) {
+  var root = path.join(SDK_PATHS.sdk, 'macos');
+  return extract(checksumVerify, filename, sdkStream, root, 2, 'Tessel build tools', bz2());
+}
+
+function extractRustlib(checksumVerify, filename, sdkStream, rustVersion) {
+  var root = path.join(SDK_PATHS.rustlib, rustVersion);
+  return extract(checksumVerify, filename, sdkStream, root, 0, 'MIPS libstd', zlib.createGunzip());
+}
+
+module.exports.getBuildConfig = () => {
+  var config = {
+    rustv: null,
+    toolchainPath: null,
+    stagingDir: null,
+    rustlibPath: null,
+    name: null,
+    path: null,
+  };
+
+  return exports.rustVersion()
+    .then(rustv => {
+      config.rustv = rustv;
+
+      return exports.checkTools();
+    })
+    .then(check => {
+      if (!check.exists) {
+        throw new Error('SDK not installed.');
+      }
+      config.stagingDir = check.path;
+
+      return exports.checkRustlib(config.rustv);
+    })
+    .then(check => {
+      if (!check.exists) {
+        throw new Error(`MIPS libstd v${config.rustv} not installed.`);
+      }
+      config.rustlibPath = check.path;
+
+      return exports.toolchainPath();
+    })
+    .then(toolchainPath => {
+      config.toolchainPath = toolchainPath;
+
+      return config;
+    });
+};
+
+// Confirms that the user has a version of rustc and cargo installed. Resolves
+// with the current rustc version, rejects if either executable is not found
+// on the system.
+module.exports.rustVersion = () => {
+  // Check that rustc exists and get its version.
+  return new Promise((resolve, reject) => {
+    var rustc = cp.spawn('rustc', ['-V']);
+    var rustcOut = [];
+    rustc.stdout.on('data', (data) => {
+      rustcOut.push(data);
+    });
+    rustc.on('error', reject);
+    rustc.on('close', (status) => {
+      var out = Buffer.concat(rustcOut).toString();
+      var matches = out.match(/^rustc\s+(\S+)/);
+      if (status !== 0 || !matches) {
+        return reject(new Error('Could not find a locally installed version of `rustc`.'));
```

Comment:
Note: don't use back ticks in strings, ie. 'Could not find a locally installed version of `rustc`.'. There is no formatting done to these strings to make those ticks useful, and they can be potentially confusing since we're using template strings throughout the code. Since we're already using single quotes for all other strings, I've been trying to convert any that I come across to double quotes for literal things. This approach is adapted directly from `git`, eg. from `git help`: 


```

'git help -a' and 'git help -g' list available subcommands and some
concept guides. See 'git help <command>' or 'git help <concept>'
to read about a specific subcommand or concept.
```